### ISSUE TITLE
Set align_flag=True yet correlation templates appear misaligned

### ISSUE THREAD

**Issue by akatav** [flatironinstitute/CaImAn#770]
Hi @epnev 
thanks for your response to my previous issue. i did set the align_flag to True but it does not seem to help, unfortunately. maybe i am missing something.

Just a bit of background: we have 6 sessions images obtained on different days from freely moving mice. there is some shift between these images obtained possibly due to head movement on different days. Individual sessions are motion corrected using caiman moco functions. There are some blood vessels that are imaged  as well and using them as landmarks, we can clearly see that the correlation images from the different sessions obtained right after the calls to register_ROIs() (we experimented with/without setting "use_opt_flow") are not aligned well to each other. We also set the shifts_opencv=False as setting it to True gives an empty xy_grid value causing error. 

For instance, these are the correlation images from subsequent sessions returned from register_ROIs: 
![corr0](https://user-images.githubusercontent.com/721570/84244123-27aa0f80-ab3e-11ea-93e4-b3de9b3d362e.png)
![corr1](https://user-images.githubusercontent.com/721570/84244134-2c6ec380-ab3e-11ea-97d8-69039cd29185.png)
![corr2](https://user-images.githubusercontent.com/721570/84244146-2f69b400-ab3e-11ea-9837-1e3c89a79f55.png)

Clearly, sessions 1 and 2 are unaligned. 
_Originally posted by @akatav in https://github.com/flatironinstitute/CaImAn/issues/764#issuecomment-641818037_

**Comment by epnev** [comment]
@akatav The aligned templates are not being saved or given as an output from the registration functions. However they're being taken into account when the registration is being performed. The goal is not to align the templates. The goal is to determine which neurons appear in multiple sessions.

**Comment by akatav** [comment]
@epnev yes, true, i agree. But i think that by setting use_opt_flow=False, template2 passed to register_ROIs is modified (after returning from tile_and_correct()). i plotted these (modified) templates (template2) after the call to register_multisession finishes and noticed that subsequent frames do not align well. 
Plotting the neurons extracted from every template onto the last template(as background), it seems that only the neurons belonging to the last template contain cell bodies. The other neurons from previous templates seem to point to empty cell bodies only. 
On the other hand, plotting neurons belonging to every session with the background set to the corresponding template for that session, the above problem does not seem to happen.

**Comment by epnev** [comment]
I think I understand what you're saying. The code does not currently return the relative shifts between the templates from different days. Essentially you want the variables `x_remap, y_remap` between all the pairs from the consecutive sessions as calculated [here](https://github.com/flatironinstitute/CaImAn/blob/master/caiman/base/rois.py#L461-L462), and you can then take the cumulative sum to obtain e.g., the mapping between the first and the last session and also apply it to the ROIs. I cannot do this at the moment, but it shouldn't be hard if you want to try it yourself. Note, that the result of the registration will not be affected at all, you'll just get access to the details of the process. 

Also in general, I've noticed leaving `use_opt_flow` to its default `True` value leads to better results.

**Comment by akatav** [comment]
@epnev thanks.So, each x_remap and y_remap inside register_ROIs() is cumulatively added up and the cumsum applied to the last template. However, this seems to result in a black background! 

I used cv2.remap() to apply the cumulative sums of x_remap and y_remap to the last template. Is this right ?

I have a few questions:

**Comment by epnev** [comment]
@akatav I had an error in my previous suggestion. What you need to accumulate over sessions are the variables `shifts_x, shifts_y` as shown [here](https://github.com/flatironinstitute/CaImAn/blob/master/caiman/base/rois.py#L458-L462), and not `x_remap, y_remap` because  `x_remap, y_remap` also contain the grid coordinates and adding those will push you out of the grid which is probably what causes the black background that you observed. Hope that helps.

> Is there a way to label each cell number in the plt.contour() ?

You can set `display_numbers=True` and that will print a number next to each contour.

**Comment by akatav** [comment]
@epnev thanks, but i think that shifts_x,shifts_y are downsized (not the same resolution as the templates) and hence remapped to the original resolution using x_grid and y_grid. is it possible to post a short code snippet to do this cumsum?

**Comment by ludovicbellier** [comment]
Hi @EricThomson,
Happy New Year! I've been encountering this same issue (i.e., too big of a shift between sessions's FoVs / template images, leading to misaligned cellsets in `register_multisession()`; also discussed [here](https://github.com/flatironinstitute/CaImAn/discussions/1173#discussion-5650388)). Found a simple fix:consisting in exposing `max_shifts` in the function's input arguments (as it's currently hardcoded as `[10, 10]`, see code here)
